datashader/glyphs/quadmesh.py

```diff
@@ -120,16 +120,132 @@ def extend(aggs, xr_ds, vt, bounds):
+        ys = xr_ds[self.y].values
+        ys = infer_interval_breaks(ys, axis=1)
+        ys = infer_interval_breaks(ys, axis=0)
+        bounds = Glyph._compute_bounds_2d(ys)
+        return self.maybe_expand_bounds(bounds)
+
+    @memoize
+    def _build_extend(self, x_mapper, y_mapper, info, append):
+        x_name = self.x
+        y_name = self.y
+
+        @ngjit
+        @self.expand_aggs_and_cols(append)
+        def _extend(plot_height, plot_width, xs, ys, *aggs_and_cols):
+            y_len, x_len, = xs.shape
+
+            for i in range(x_len - 1):
+                for j in range(y_len - 1):
+
+                    # Extract quad vertices
+                    x1 = xs[j, i]
+                    x2 = xs[j, i + 1]
+                    x3 = xs[j + 1, i + 1]
+                    x4 = xs[j + 1, i]
+
+                    y1 = ys[j, i]
+                    y2 = ys[j, i + 1]
+                    y3 = ys[j + 1, i + 1]
+                    y4 = ys[j + 1, i]
+
+                    # Compute the rectilinear bounding box around the quad
+                    xmin = max(min(x1, x2, x3, x4), 0)
+                    xmax = min(max(x1, x2, x3, x4), plot_width - 1)
+                    ymin = max(min(y1, y2, y3, y4), 0)
+                    ymax = min(max(y1, y2, y3, y4), plot_height - 1)
+
+                    # Make sure single pixel quads are represented
+                    if xmin == xmax:
+                        xmax += 1
+
+                    if ymin == ymax:
+                        ymax += 1
+
+                    in_quad = []
+                    for xi in range(xmin, xmax):
+                        for yi in range(ymin, ymax):
+                            if point_in_quad(
+                                    x1, x2, x3, x4, y1, y2, y3, y4, xi, yi):
+                                append(j, i, xi, yi, *aggs_and_cols)
+                                in_quad.append((xi, y1))
+
+        def extend(aggs, xr_ds, vt, bounds):
+            # Convert from bin centers to interval edges
+            x_breaks = xr_ds[x_name].values
+            x_breaks = infer_interval_breaks(x_breaks, axis=1)
+            x_breaks = infer_interval_breaks(x_breaks, axis=0)
+
+            y_breaks = xr_ds[y_name].values
+            y_breaks = infer_interval_breaks(y_breaks, axis=1)
+            y_breaks = infer_interval_breaks(y_breaks, axis=0)
+
+            # Scale x and y vertices into integer canvas coordinates
+            x0, x1, y0, y1 = bounds
+            xspan = x1 - x0
+            yspan = y1 - y0
+            xscaled = (x_mapper(x_breaks) - x0) / xspan
+            yscaled = (y_mapper(y_breaks) - y0) / yspan
+
+            plot_height, plot_width = aggs[0].shape[:2]
+
+            xs = (xscaled * plot_width).astype(int)
+            ys = (yscaled * plot_height).astype(int)
+
+            # Question: Should we try to compute a slice of xs and ys that
+            # eliminates rows and columns of quads that are all outside the
+            # viewport?
```

Comment:
I tried this but for many real world meshes this wasn't all that effective to reduce the array size. May still be worth it though.